testing/trino-plugin-reader/src/main/java/io/trino/server/PluginReader.java

```diff
@@ -0,0 +1,255 @@
+import static java.util.Objects.requireNonNullElse;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+@Command(name = "modulesToConnectors", mixinStandardHelpOptions = true,
+        description = "Maps Trino plugin modules to connectors they provide and filters them using an impacted modules list.")
+public class PluginReader
+        implements Callable<Integer>
+{
+    private static final Logger log = Logger.get(PluginReader.class);
+
+    @Option(names = {"-i", "--impacted-modules"}, description = "Impacted modules file generated by the gitflow-incremental-builder (GIB) Maven plugin")
+    private Optional<File> impactedModulesFile;
+
+    @Option(names = {"-p", "--plugin-dir"}, description = "Trino plugin directory")
+    private File pluginDir = new File("plugin");
+
+    @Option(names = {"-r", "--root-pom"}, description = "Trino root module pom.xml")
+    private File rootPom = new File("pom.xml");
+
+    public static void main(String... args)
+    {
+        int exitCode = new CommandLine(new PluginReader()).execute(args);
+        System.exit(exitCode);
+    }
+
+    @Override
+    public Integer call()
+    {
+        Optional<List<String>> impactedModules = Optional.empty();
+        if (impactedModulesFile.isPresent()) {
+            impactedModules = readImpactedModules(impactedModulesFile.get());
+            if (impactedModules.isEmpty()) {
+                return 1;
+            }
+        }
+        Map<String, String> modulesToPlugins = mapModulesToPlugins(rootPom);
+        Map<String, Plugin> plugins = loadPlugins(pluginDir).stream()
+                .collect(toMap(plugin -> plugin.getClass().getName(), identity()));
+        Stream<Map.Entry<String, String>> modulesStream = requireNonNull(modulesToPlugins).entrySet().stream();
+        if (impactedModules.isPresent()) {
+            List<String> nonPluginModules = impactedModules.get().stream().filter(module -> !modulesToPlugins.containsKey(module)).collect(Collectors.toList());
+            if (nonPluginModules.size() != 0) {
+                log.warn("Impacted modules list includes non-plugin modules, ignoring it: %s", nonPluginModules);
+            }
+            else {
+                List<String> finalImpactedModules = impactedModules.get();
+                modulesStream = modulesStream.filter(entry -> finalImpactedModules.contains(entry.getKey()));
+            }
+        }
+
+        modulesStream.forEach(entry -> {
+            if (!plugins.containsKey(entry.getValue())) {
+                log.warn("Plugin without any connectors: %s", entry.getValue());
+                return;
+            }
+            printPluginFeatures(plugins.get(entry.getValue()));
+        });
+        return 0;
+    }
+
+    private static void printPluginFeatures(Plugin plugin)
+    {
+        plugin.getConnectorFactories().forEach(factory -> System.out.println("connector:" + factory.getName()));
```

Comment:
Is it possible to add test that would detect when new methods have been added to `Plugin` via reflection?

(Not required but useful if we want to maintain this class).